vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -489,35 +644,141 @@ static void run_dijkstra(int start_node_ind,
+            for (int y = 0; y < SAMPLE_GRID_SIZE; y++) {
+                for (int x = 0; x < SAMPLE_GRID_SIZE; x++) {
+                    auto& point = grid.point[y][x];
+                    if (!point.samples.empty()) {
+                        point.order = point.samples[0];
+                        sample_points.push_back(point);
+                    }
+                }
             }
+        }
+    }
+
+    // sort by VPR coordinate
+    std::sort(sample_points.begin(), sample_points.end(),
+              [](const SamplePoint& a, const SamplePoint& b) {
+                  return a.order < b.order;
+              });
 
-            int delta_x = ssize_t(from_canonical_loc->first) - ssize_t(box_location.first);
-            int delta_y = ssize_t(from_canonical_loc->second) - ssize_t(box_location.second);
+    /* free previous delay map and allocate new one */
+    auto& device_ctx = g_vpr_ctx.device();
+    cost_map_.set_counts(segment_inf.size(),
+                         device_ctx.connection_boxes.num_connection_box_types());
+
+    VTR_ASSERT(REPRESENTATIVE_ENTRY_METHOD == SMALLEST);
+    RoutingCosts all_costs;
+
+    /* run Dijkstra's algorithm for each segment type & channel type combination */
+#if defined(VPR_USE_TBB)
+    tbb::mutex all_costs_mutex;
+    tbb::parallel_for_each(sample_points, [&](const SamplePoint& point) {
+#else
+    for (const auto& point : sample_points) {
+#endif
+        float max_cost = 0.f;
 
-            routing_cost_map->push_back(std::make_tuple(
-                std::make_pair(start_node_ind, node_ind),
-                std::make_pair(delta_x, delta_y),
-                Cost_Entry(
-                    current.delay,
-                    current.congestion_upstream),
-                box_id));
+        // holds the cost entries for a run
+        RoutingCosts costs;
+
+        // a cache to avoid hammering the RoutingCosts map, since lookups will be dominated by a few keys
+        // must be consistent with `costs` i.e. any entry in the cache should also be in `costs`
+        // NOTE: this is used as a write-through cache, maybe try write-back
+        SimpleCache<CompressedRoutingCostKey, float, DIJKSTRA_CACHE_SIZE> cache;
+
+        for (auto node_ind : point.samples) {
+            max_cost = std::max(max_cost, run_dijkstra(node_ind, &costs, &cache, COST_LIMIT));
         }
 
-        expand_dijkstra_neighbours(current, node_visited_costs, node_expanded, pq);
-        node_expanded[node_ind] = true;
+#if defined(VPR_USE_TBB)
+        all_costs_mutex.lock();
+#endif
+
+        VTR_LOG("Expanded sample point (%d, %d) %e miss %g\n",
+                point.location.x(), point.location.y(), max_cost, cache.miss_ratio());
+
+        // combine the cost map from this run with the final cost maps for each segment
+        for (const auto& cost : costs) {
+            const auto& key = cost.first;
+            const auto& val = cost.second;
+            const auto& x = all_costs.find(key);
```

Comment:
Another instance where `find` should be `insert` to avoid the double hash.

Reply:
Fixed.